core/optaplanner-core-impl/src/main/java/org/optaplanner/core/api/solver/change/ProblemChange.java

```diff
@@ -18,18 +18,63 @@
 
 import org.optaplanner.core.api.domain.entity.PlanningEntity;
 import org.optaplanner.core.api.domain.solution.PlanningSolution;
-import org.optaplanner.core.api.score.Score;
 import org.optaplanner.core.api.solver.Solver;
 
 /**
- * A ProblemChange represents a change in 1 or more {@link PlanningEntity planning entities} or problem facts
+ * A ProblemChange represents a change in one or more {@link PlanningEntity planning entities} or problem facts
  * of a {@link PlanningSolution}.
- * Problem facts used by a {@link Solver} must not be changed while it is solving,
- * but by scheduling this command to the {@link Solver}, you can change them when the time is right.
+ * <p>
+ * The {@link Solver} checks the presence of waiting problem changes after every
+ * {@link org.optaplanner.core.impl.heuristic.move.Move} evaluation. If there are waiting problem changes,
+ * the {@link Solver}:
+ * <ol>
+ *  <li>clones the last {@link PlanningSolution best solution} and sets the clone
+ *  as the new {@link PlanningSolution working solution}</li>
+ *  <li>applies every problem change keeping the order in which problem changes have been submitted;
+ *  if {@link ProblemChange#usesShadowVariables()} returns {@code true},
+ *  {@link org.optaplanner.core.api.domain.variable.VariableListener variable listeners} are triggered before the
+ *  problem change is applied</li>
+ *  <li>triggers {@link org.optaplanner.core.api.domain.variable.VariableListener variable listeners} to guarantee
+ *  consistency of the {@link PlanningSolution updated working solution}</li>
+ *  <li>calculates the score and makes the {@link PlanningSolution updated working solution}
+ *  the new {@link PlanningSolution best solution}; note that this {@link PlanningSolution solution} is not published
+ *  via the {@link org.optaplanner.core.api.solver.event.BestSolutionChangedEvent}</li>
```

Comment:
Very nice explanation! I'd maybe add an extra note here, explaining why the new BKS is not published.
(Because it may not even be initialized.)